### PR TITLE
Get SQL service objectives by location

### PR DESCRIPTION
**Pull request by jaredmoo**
## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (Microsoft internal partners only)
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/help-generation.md)

### CODE REVIEW

File: src/Sql/Sql/ServiceObjective/Cmdlet/GetAzureSqlServerServiceObjective.cs [hunk 1]
```diff
@@ -44,7 +45,11 @@ protected override IEnumerable<AzureSqlServerServiceObjectiveModel> GetEntity()
         {
             ICollection<AzureSqlServerServiceObjectiveModel> results = null;
 
-            if (this.MyInvocation.BoundParameters.ContainsKey("ServiceObjectiveName") && !WildcardPattern.ContainsWildcardCharacters(ServiceObjectiveName))
+            if (this.ParameterSetName == ByLocationNameParameterSet)
```
Comment (akromm) [inline src/Sql/Sql/ServiceObjective/Cmdlet/GetAzureSqlServerServiceObjective.cs]: should filtering by SLO name also be supported with LocationName param set?

Comment (jaredmoo) [reply src/Sql/Sql/ServiceObjective/Cmdlet/GetAzureSqlServerServiceObjective.cs]: Yes, done

File: src/Sql/Sql/ServiceObjective/Cmdlet/AzureSqlServerServiceObjectiveCmdletBase.cs [hunk 2]
```diff
@@ -27,17 +27,46 @@ namespace Microsoft.Azure.Commands.Sql.ServiceObjective.Cmdlet
...
         public string ServerName { get; set; }
 
+        /// <summary>
+        /// Gets or sets the name of the database server to use.
+        /// </summary>
+        [Parameter(Mandatory = true,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "The name of the Location for which to get the service objectives.",
+            ParameterSetName = ByLocationNameParameterSet)]
+        [LocationCompleter("Microsoft.Sql/locations/capabilities")]
```
[the remark this answers is not in the document]
Comment (jaredmoo) [reply src/Sql/Sql/ServiceObjective/Cmdlet/AzureSqlServerServiceObjectiveCmdletBase.cs]: I copied this from get-capabilities command and it seems to actually work.

[the remark this answers is not in the document]
Comment (jaredmoo) [reply src/Sql/Sql/ServiceObjective/Cmdlet/AzureSqlServerServiceObjectiveCmdletBase.cs]: https://github.com/Azure/azure-powershell-common/blob/master/src/ResourceManager/Version2016_09_01/ArgumentCompleters/LocationCompleter.cs#L114

File: src/Sql/Sql/ChangeLog.md [hunk 3]
```diff
@@ -22,6 +22,7 @@
 * Updated cmdlets with plural nouns to singular, and deprecated plural names.
 * Enhanced process of multiple columns classification.
 * Include sku properties (sku name, family, capacity) in response from Get-AzSqlServerServiceObjective and format as table by default.
+* Ability to Get-AzSqlServerServiceObjective by location without needing a server to already exist there.
```
Comment (akromm) [inline src/Sql/Sql/ChangeLog.md]: ... without needing a preexisting server in the region.

File: src/Sql/Sql/ServiceObjective/Cmdlet/AzureSqlServerServiceObjectiveCmdletBase.cs [hunk 4]
```diff
@@ -27,17 +27,46 @@ namespace Microsoft.Azure.Commands.Sql.ServiceObjective.Cmdlet
...
+        public override string ResourceGroupName { get; set; }
+
         /// <summary>
         /// Gets or sets the name of the database server to use.
         /// </summary>
         [Parameter(Mandatory = true,
             ValueFromPipelineByPropertyName = true,
             Position = 1,
-            HelpMessage = "SQL Database server name.")]
+            HelpMessage = "SQL Database server name.",
```
Comment (akromm) [inline src/Sql/Sql/ServiceObjective/Cmdlet/AzureSqlServerServiceObjectiveCmdletBase.cs]: Azure SQL Server name.

File: src/Sql/Sql/ServiceObjective/Cmdlet/AzureSqlServerServiceObjectiveCmdletBase.cs [hunk 5]
```diff
@@ -27,17 +27,46 @@ namespace Microsoft.Azure.Commands.Sql.ServiceObjective.Cmdlet
...
+        /// <summary>
+        /// Gets or sets the name of the database server to use.
+        /// </summary>
+        [Parameter(Mandatory = true,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "The name of the Location for which to get the service objectives.",
+            ParameterSetName = ByLocationNameParameterSet)]
+        [LocationCompleter("Microsoft.Sql/locations/capabilities")]
+        [ValidateNotNullOrEmpty]
+        public string LocationName { get; set; }
```
Comment (maddieclayton) [inline src/Sql/Sql/ServiceObjective/Cmdlet/AzureSqlServerServiceObjectiveCmdletBase.cs]: Generally across the modules we would call this "Location". Please update - if you really think that "LocationName" provides clarity you can add it as an alias.

File: src/Sql/Sql/ServiceObjective/Cmdlet/GetAzureSqlServerServiceObjective.cs [hunk 6]
```diff
@@ -42,21 +47,19 @@ public class GetAzureSqlServerServiceObjective : AzureSqlServerServiceObjectiveC
         /// <returns>A single server</returns>
         protected override IEnumerable<AzureSqlServerServiceObjectiveModel> GetEntity()
         {
-            ICollection<AzureSqlServerServiceObjectiveModel> results = null;
-
-            if (this.MyInvocation.BoundParameters.ContainsKey("ServiceObjectiveName") && !WildcardPattern.ContainsWildcardCharacters(ServiceObjectiveName))
```
Comment (maddieclayton) [inline src/Sql/Sql/ServiceObjective/Cmdlet/GetAzureSqlServerServiceObjective.cs]: Why are you implementing your own wildcard support? Please just keep the current implementation.

File: src/Sql/Sql/ServiceObjective/Cmdlet/AzureSqlServerServiceObjectiveCmdletBase.cs (hunk 5 above)
Comment (maddieclayton) [reply src/Sql/Sql/ServiceObjective/Cmdlet/AzureSqlServerServiceObjectiveCmdletBase.cs]: Be sure to regen the docs after making this change

File: src/Sql/Sql/ServiceObjective/Cmdlet/GetAzureSqlServerServiceObjective.cs (hunk 6 above)
Comment (jaredmoo) [reply src/Sql/Sql/ServiceObjective/Cmdlet/GetAzureSqlServerServiceObjective.cs]: The returned objects don't have "Id" or "ResourceId" properties, so the current implementation doesn't work.

File: src/Sql/Sql/ServiceObjective/Cmdlet/AzureSqlServerServiceObjectiveCmdletBase.cs (hunk 5 above)
Comment (jaredmoo) [reply src/Sql/Sql/ServiceObjective/Cmdlet/AzureSqlServerServiceObjectiveCmdletBase.cs]: Can do, but I copied this from `Get-AzSqlCapability` which has -LocationName. I guess we should also fix that.

File: src/Sql/Sql/ServiceObjective/Cmdlet/GetAzureSqlServerServiceObjective.cs (hunk 6 above)
Comment (maddieclayton) [reply src/Sql/Sql/ServiceObjective/Cmdlet/GetAzureSqlServerServiceObjective.cs]: Please run the "Get"  if the Service objective does not contain wildcards, as the list could take much longer/in the future have a different set of properties.

Comment (jaredmoo) [reply src/Sql/Sql/ServiceObjective/Cmdlet/GetAzureSqlServerServiceObjective.cs]: SQL capabilities is a really weird API. It only has "List". https://docs.microsoft.com/en-us/rest/api/sql/capabilities/listbylocation

**Comment by jaredmoo** [comment]
Thank you for the reviews @maddieclayton and @akromm !

**Pull request merged**